### PR TITLE
feat: B3 support

### LINKED ISSUE #196: B3 context propagation
This not spec'd yet, but will likely be needed for users adopting OTel and transitioning to W3C trace context. See: https://github.com/openzipkin/b3-propagation and https://github.com/open-telemetry/opentelemetry-specification/issues/8.

Python and JS have B3 support.

### PR DESCRIPTION
**Pull request by mwear**
Fixes #196.

This PR adds injectors and extractors for b3 context propagation. There are injectors for both the B3 Multi and B3 Single encodings. Users can configure as few or as many injection or extraction formats as their application needs at the SDK level. The b3 implementation is based on https://github.com/open-telemetry/opentelemetry-specification/pull/1045, which describes the current state of b3 propagation in OpenTelemetry.

### CODE REVIEW

File: propagator/b3/lib/opentelemetry/propagator/b3/single/text_map_injector.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      # Namespace for OpenTelemetry b3 single header encoding
+      module Single
+        # Injects context into carriers using the b3 single header format
+        class TextMapInjector
+          include Context::Propagation::DefaultSetter
+
+          # Returns a new TextMapInjector that extracts b3 context using the
+          # specified header keys
+          #
+          # @param [String] b3_key The traceparent header key used in the carrier
```
Comment (fbogsany) [inline propagator/b3/lib/opentelemetry/propagator/b3/single/text_map_injector.rb:20]:
```suggestion
          # @param [String] b3_key The b3 header key used in the carrier
```

File: propagator/b3/lib/opentelemetry/propagator/b3/single/text_map_injector.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+          # @param [Context] context The active Context.
+          # @param [optional Callable] setter An optional callable that takes a carrier and a key and
+          #   a value and assigns the key-value pair in the carrier. If omitted the default setter
+          #   will be used which expects the carrier to respond to [] and []=.
+          # @yield [Carrier, String, String] if an optional setter is provided, inject will yield
+          #   carrier, header key, header value to the setter.
+          # @return [Object] the carrier with context injected
+          def inject(carrier, context, &setter)
+            span_context = Trace.current_span(context)&.context
+            return unless span_context&.valid?
```
Comment (fbogsany) [inline propagator/b3/lib/opentelemetry/propagator/b3/single/text_map_injector.rb:37]:
```suggestion
            span_context = Trace.current_span(context).context
            return unless span_context.valid?
```
`current_span` should always return a non-`nil` span.

File: propagator/b3/lib/opentelemetry/propagator/b3/single/text_map_injector.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+                               'd'
+                             elsif span_context.trace_flags.sampled?
+                               '1'
+                             else
+                               '0'
+                             end
+
+            b3_value = "#{span_context.hex_trace_id}-#{span_context.hex_span_id}-#{sampling_state}"
+
+            setter ||= DEFAULT_SETTER
```
Comment (fbogsany) [inline propagator/b3/lib/opentelemetry/propagator/b3/single/text_map_injector.rb:49]:
```suggestion
            setter ||= default_setter
```

File: propagator/b3/lib/opentelemetry/propagator/b3/multi/text_map_injector.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+          #   a value and assigns the key-value pair in the carrier. If omitted the default setter
+          #   will be used which expects the carrier to respond to [] and []=.
+          # @yield [Carrier, String, String] if an optional setter is provided, inject will yield
+          #   carrier, header key, header value to the setter.
+          # @return [Object] the carrier with context injected
+          def inject(carrier, context, &setter)
+            span_context = Trace.current_span(context)&.context
+            return unless span_context&.valid?
+
+            setter ||= DEFAULT_SETTER
```
Comment (fbogsany) [inline propagator/b3/lib/opentelemetry/propagator/b3/multi/text_map_injector.rb:48]:
```suggestion
            setter ||= default_setter
```

File: propagator/b3/lib/opentelemetry/propagator/b3/multi/text_map_injector.rb [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+          # @param [Context] context The active Context.
+          # @param [optional Callable] setter An optional callable that takes a carrier and a key and
+          #   a value and assigns the key-value pair in the carrier. If omitted the default setter
+          #   will be used which expects the carrier to respond to [] and []=.
+          # @yield [Carrier, String, String] if an optional setter is provided, inject will yield
+          #   carrier, header key, header value to the setter.
+          # @return [Object] the carrier with context injected
+          def inject(carrier, context, &setter)
+            span_context = Trace.current_span(context)&.context
+            return unless span_context&.valid?
```
Comment (fbogsany) [inline propagator/b3/lib/opentelemetry/propagator/b3/multi/text_map_injector.rb:46]:
```suggestion
            span_context = Trace.current_span(context).context
            return unless span_context.valid?
```

File: propagator/b3/lib/opentelemetry/propagator/b3/multi/text_map_extractor.rb [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+      module Multi
+        # Extracts context from carriers in the b3 single header format
+        class TextMapExtractor
+          include Context::Propagation::DefaultGetter
+
+          B3_TRACE_ID_REGEX = /\A(?:[0-9a-f]{16}){1,2}\z/.freeze
+          B3_SPAN_ID_REGEX = /\A[0-9a-f]{16}\z/.freeze
+          SAMPLED_VALUES = %w[1 true].freeze
+          DEBUG_FLAG = '1'
+
```
Comment (fbogsany) [inline propagator/b3/lib/opentelemetry/propagator/b3/multi/text_map_extractor.rb:28]:
```suggestion
          private_constant :B3_TRACE_ID_REGEX, :B3_SPAN_ID_REGEX, :SAMPLED_VALUES, :DEBUG_FLAG

```

File: propagator/b3/lib/opentelemetry/propagator/b3.rb [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        context.value(DEBUG_CONTEXT_KEY)
+      end
+
+      # @api private
+      # Convert an id from a hex encoded string to byte array, optionally left
+      # padding to the correct length. Assumes the input id has already been
+      # validated to be 16 or 32 characters in length.
+      def to_trace_id(hex_id)
+        hex_id = "#{PADDING}#{hex_id}" unless hex_id.length == 32
+        Array(hex_id).pack('H*')
```
Comment (fbogsany) [inline propagator/b3/lib/opentelemetry/propagator/b3.rb:42]:
```suggestion
        if hex_id.length == 32
          Array(hex_id).pack('H*')
        else
          [0, hex_id].pack('qH*')
        end
```

Comment (mwear) [reply propagator/b3/lib/opentelemetry/propagator/b3.rb]: TIL. Thanks!

**Review by fbogsany — APPROVED** [review]

**Pull request merged**